actors/evm/src/interpreter/precompiles.rs

```diff
@@ -179,28 +382,140 @@ fn lookup_address<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+/// for exit_code:
+/// - negative values are system errors
+/// - positive are user errors (from the called actor)
+/// - 0 is success
+pub fn call_actor<RT: Runtime>(rt: &RT, input: &[u8], ctx: PrecompileContext) -> PrecompileResult {
+    // TODO Until we support readonly (static) calls at the fvm level, we disallow callactor
+    //      when in static mode as it is sticky and there are no guarantee of preserving the
+    //      static invariant
+    if ctx.is_static {
+        return Err(PrecompileError::CallActorError(StatusCode::StaticModeViolation));
+    }
+
+    // ----- Input Parameters -------
+
+    let mut input_params = U256Reader::new(input);
+
+    let method: u64 = input_params.next_param_padded()?;
+    let codec: u64 = input_params.next_param_padded()?;
+    // TODO only CBOR for now
+    if codec != fvm_ipld_encoding::DAG_CBOR {
+        return Err(PrecompileError::InvalidInput);
+    }
+
+    let address_size = input_params.next_param_padded::<u32>()? as usize;
+    let send_data_size = input_params.next_param_padded::<u32>()? as usize;
+
+    // ------ Begin Call -------
+
+    let result = {
+        // REMOVEME: closes https://github.com/filecoin-project/ref-fvm/issues/1018
+
+        let start = input_params.remaining_slice();
+        let bytes = read_right_pad(start, send_data_size + address_size);
+
+        let input_data = &bytes[..send_data_size];
+        let address = &bytes[send_data_size..address_size];
+        let address = Address::from_bytes(address).map_err(|_| PrecompileError::InvalidInput)?;
+
+        // TODO passing underlying gas into send when implemented in FVM
+        rt.send(
+            &address,
+            method,
+            RawBytes::from(input_data.to_vec()),
+            TokenAmount::from(&ctx.value),
+        )
+    };
+
+    // ------ Build Output -------
+
+    let output = {
+        // negative values are syscall errors
+        // positive values are user/actor errors
+        // success is 0
+        let (exit_code, data) = match result {
+            Err(ae) => {
+                // TODO handle revert
+                // put error number from call into revert
+                let exit_code = U256::from(ae.exit_code().value());
+                let exit_code = if ae.exit_code().is_system_error() {
+                    exit_code.i256_neg()
+                } else {
+                    U256::zero()
+                };
+
+                // no return only exit code
+                (exit_code, RawBytes::default())
+            }
+            Ok(ret) => (U256::zero(), ret),
+        };
+
+        let num_output_params = 3;
+
+        // codec of return data
+        // TODO hardcoded to CBOR for now
+        let codec = U256::from(fvm_ipld_encoding::DAG_CBOR);
+        let offset = U256::from(num_output_params * 32);
+        let size = U256::from(data.len());
+
+        let mut output = Vec::with_capacity(num_output_params * 32 + data.len());
+        output.extend_from_slice(
```

Comment:
Just call `extend_from_slice` multiple times (concat will copy and allocate multiple times).

Reply:
`let s: &[u8] = &[exit_code.to_bytes(), codec.to_bytes(), offset.to_bytes(), size.to_bytes()].concat();`

should work just fine? this would result in just the copy into the vec

Reply:
ah no this makes a vec underneath..

Reply:
Unfortunately, yes. An alternative is to turn them into iterators then join the iterators... but that can get real slow unless the compiler figures out what you're doing.